File: maxtemperature-monoid-folding/src/MaxTemperatureMonoidFolding/MaxTemperature.cs
namespace MaxTemperatureMonoidFolding;

public class MaxTemperature : IMonoid<(int, int)> {
    public (int, int) Zero() {
        return (0, 0);
    }

    public bool IsZero((int, int) v) {
        (int time, int value) = v;
        return time.Equals(0) && value.Equals(0);
    }

    public (int, int) Op((int, int) a, (int, int) b) {
        if (IsZero(a)) {
            return b;
        }
        if (IsZero(b)) {
            return a;
        }

        (var _, var aValue) = a;
        (var _, var bValue) = b;
        return aValue > bValue
            ? a
            : b;
    }
}

File: monoid-folding-test-typescript/src/MonoidFoldingTest/MaxTemperature.ts
import { IMonoid } from "./IMonoid";

export class MaxTemperature implements IMonoid<[number, number]> {
    zero(): [number, number] {
        return [0, 0];
    }

    isZero(v: [number, number]): boolean {
        const [time, value] = v;
        return time === 0 && value === 0;
    }

    op(a: [number, number], b: [number, number]): [number, number] {
        if (this.isZero(a)) {
            return b;
        }
        if (this.isZero(b)) {
            return a;
        }

        const [, aValue] = a;
        const [, bValue] = b;

        return aValue > bValue
            ? a
            : b;
    }
}
